apps/astarte_data_updater_plant/lib/astarte_data_updater_plant/data_updater/state.ex

```diff
@@ -17,35 +17,91 @@
 #
 
 defmodule Astarte.DataUpdaterPlant.DataUpdater.State do
-  defstruct [
-    :realm,
-    :device_id,
-    :message_tracker,
-    :introspection,
-    :groups,
-    :interfaces,
-    :interface_ids_to_name,
-    :interfaces_by_expiry,
-    :mappings,
-    :paths_cache,
-    :device_triggers,
-    :data_triggers,
-    :volatile_triggers,
-    :introspection_triggers,
-    :connected,
-    :total_received_msgs,
-    :total_received_bytes,
-    :initial_interface_exchanged_bytes,
-    :initial_interface_exchanged_msgs,
-    :interface_exchanged_bytes,
-    :interface_exchanged_msgs,
-    :last_seen_message,
-    :last_device_triggers_refresh,
-    :last_groups_refresh,
-    :datastream_maximum_storage_retention,
-    :trigger_id_to_policy_name,
-    :discard_messages,
-    :last_deletion_in_progress_refresh,
-    :last_datastream_maximum_retention_refresh
-  ]
+  use TypedStruct
+
+  @type interface_name :: String.t()
+  @type policy_name :: String.t()
+  @type group_name :: String.t()
+
+  @type descriptors_by_name :: %{interface_name => InterfaceDescriptor.t()}
+
+  @type volatile_trigger_key :: {binary(), integer()}
+
+  @type volatile_trigger_value ::
+          {{:data_trigger, Astarte.Core.Triggers.DataTrigger.t()},
+           Astarte.Core.Triggers.SimpleTriggersProtobuf.AMQPTriggerTarget.t()}
+
+  @type volatile_triggers_type :: [{volatile_trigger_key(), volatile_trigger_value()}]
+
+  @type mappings_map :: %{binary() => Astarte.Core.Mapping.t()}
+
+  @type interface_exchanged_msgs_map :: %{{String.t(), non_neg_integer()} => non_neg_integer()}
+  @type interface_exchanged_bytes_map :: %{{String.t(), non_neg_integer()} => non_neg_integer()}
+  @type paths_cache_type ::
+          {non_neg_integer(),
+           %{
+             {String.t(), String.t()} =>
+               {Astarte.DataUpdaterPlant.DataUpdater.CachedPath.t(), nil | term()}
+           }}
```

Comment:
nit: this could be added to `Astarte.DataUpdaterPlant.DataUpdater.Cache`:

```elixir
  @typep cache_size :: non_neg_integer()
  @typep expiry :: integer() | nil

  @type t() :: {cache_size(), %{}}
  @type t(key, value) :: {cache_size(), %{key => {value, expiry()}}}
```

so you can define your types

```suggestion
  @type path :: String.t()
  @type paths_cache_key :: {interface_name, path}

  @type paths_cache_type :: Cache.t(paths_cache_key, CachedPath.t())
```